### ISSUE TITLE
multiprocessing.get fragile on Windows

### ISSUE THREAD

**Issue by pitrou** [dask/dask#1652]
`multiprocessing.get` uses a multiprocessing Manager object. That object may unpickle objects from several threads at once; unfortunately, in some cases this will fail importing `numpy` if it wasn't already imported (concurrent circular imports may indeed expose the module in an incomplete state). The symptom may be:

```
E                       dask.async.RemoteError: 
E                       ---------------------------------------------------------------------------
E                       Traceback (most recent call last):
E                         File "C:\Miniconda35-x64\envs\testenv\lib\multiprocessing\managers.py", line 228, in serve_client
E                           request = recv()
E                         File "C:\Miniconda35-x64\envs\testenv\lib\multiprocessing\connection.py", line 251, in recv
E                           return ForkingPickler.loads(buf.getbuffer())
E                       AttributeError: Can't get attribute 'ndarray' on <module 'numpy' from 'C:\\Miniconda35-x64\\envs\\testenv\\lib\\site-packages\\numpy\\__init__.py'>
E                       ---------------------------------------------------------------------------
E                       
E                       Traceback
E                       ---------
E                         File "C:\projects\dask\dask\async.py", line 273, in execute_task
E                           queue.put(result)
E                         File "<string>", line 2, in put
E                         File "C:\Miniconda35-x64\envs\testenv\lib\multiprocessing\managers.py", line 732, in _callmethod
E                           raise convert_to_error(kind, result)
dask\async.py:481: RemoteError
```
(in `dask/array/tests/test_random.py::test_concurrency`, see https://ci.appveyor.com/project/dask-ci/dask/build/1.0.1)

**Comment by pitrou** [comment]
The only reasonable solution would be to do without a Manager object (which may also speed up `multiprocessing.get`, since the Manager running in a separate process may incur additional serialization costs). A hack doing so is at https://gist.github.com/pitrou/72e186745c7166197d13d581f851f67c, but is not a very pretty solution.

**Comment by mrocklin** [comment]
@pitrou offline you mentioned that we could get around this by using a callback from the `apply_async` function.  Can you say a bit more about how this would work?  It has been a while since I looked deeply at `dask/async.py`, so please excuse my ignorance here.

The queue currently in `dask/async.py::get_async` currently serves two purposes:

1.  It communicates data back from the worker if necessary (useful for multiprocessing only)
2.  It serves as a point of coordination between the workers and the main thread/process

Presumably we can use the `.get()` method on the async result to communicate data back.  What would we use as a point of coordination between workers?  Where would we call the `.get()` method on the async results?

**Comment by mrocklin** [comment]
Ah!  `apply_async` itself takes a `callback=` keyword.  I see.

**Comment by mrocklin** [comment]
I have a slight preference to keep results coming in through a central queue that we wait on in the main loop.  I suppose the callback could just put the results on a local Queue though as is done in your hack above.  This seems reasonable to me.

Where do existing pool objects run callbacks?  In the multiprocessing case is the callback run in the main process?

**Comment by pitrou** [comment]
The callback is run in the collecting process, but in a separate thread. So, yes, a Queue object is still necessary anyway.

**Comment by pitrou** [comment]
Closing since this is fixed by #1653.
